File: test/test_protocol.py
import struct

import pytest

from netft_driver.protocol import Command, encode_request


@pytest.mark.parametrize(
    "command,value",
    [
        (Command.STOP_STREAMING, 0x0000),
        (Command.START_REALTIME, 0x0002),
        (Command.START_BUFFERED, 0x0003),
        (Command.RESET_CONDITION_LATCH, 0x0041),
        (Command.SET_SOFTWARE_BIAS, 0x0042),
    ],
)
def test_encode_request_uses_network_byte_order(command, value):
    assert encode_request(command, 7) == struct.pack(">HHI", 0x1234, value, 7)


@pytest.mark.parametrize("sample_count", [-1, 0x1_0000_0000])
def test_encode_request_rejects_sample_count_outside_uint32(sample_count):
    with pytest.raises(ValueError, match="sample_count"):
        encode_request(Command.START_REALTIME, sample_count)


from netft_driver.protocol import ProtocolError, decode_record


def test_decode_record_preserves_unsigned_headers_and_signed_axes():
    payload = struct.pack(
        ">III6i",
        0xFFFFFFFF,
        0x80000000,
        0x80020000,
        -1,
        2,
        -3,
        4,
        -5,
        6,
    )

    record = decode_record(payload)

    assert record.rdt_sequence == 0xFFFFFFFF
    assert record.ft_sequence == 0x80000000
    assert record.status == 0x80020000
    assert (record.fx, record.fy, record.fz) == (-1, 2, -3)
    assert (record.tx, record.ty, record.tz) == (4, -5, 6)


@pytest.mark.parametrize("size", [0, 7, 35, 37, 72])
def test_decode_record_rejects_non_36_byte_datagrams(size):
    with pytest.raises(ProtocolError, match="exactly 36 bytes"):
        decode_record(bytes(size))

File: test/cpp/test_protocol.cpp
#include <gtest/gtest.h>

#include <array>
#include <cstdint>
#include <fstream>
#include <iterator>
#include <limits>
#include <string>
#include <vector>

#include "netft_driver/protocol.hpp"

namespace {

using netft_driver::Command;
using netft_driver::ProtocolError;
using netft_driver::decode_record;
using netft_driver::encode_request;

TEST(Protocol, EncodesRequestsInNetworkByteOrder)
{
  EXPECT_EQ(encode_request(Command::StartRealtime, 0),
    (std::array<std::uint8_t, 8>{0x12, 0x34, 0x00, 0x02, 0, 0, 0, 0}));
  EXPECT_EQ(encode_request(Command::StopStreaming, 7),
    (std::array<std::uint8_t, 8>{0x12, 0x34, 0x00, 0x00, 0, 0, 0, 7}));
  EXPECT_EQ(encode_request(Command::StartBuffered, 7),
    (std::array<std::uint8_t, 8>{0x12, 0x34, 0x00, 0x03, 0, 0, 0, 7}));
  EXPECT_EQ(encode_request(Command::ResetConditionLatch, 7),
    (std::array<std::uint8_t, 8>{0x12, 0x34, 0x00, 0x41, 0, 0, 0, 7}));
  EXPECT_EQ(encode_request(Command::SetSoftwareBias, 7),
    (std::array<std::uint8_t, 8>{0x12, 0x34, 0x00, 0x42, 0, 0, 0, 7}));
}

TEST(Protocol, DecodesUnsignedHeadersAndSignedAxes)
{
  const std::array<std::uint8_t, 36> payload{
    0xff, 0xff, 0xff, 0xff, 0x80, 0, 0, 0, 0x80, 0x02, 0, 0,
    0xff, 0xff, 0xff, 0xff, 0, 0, 0, 2, 0xff, 0xff, 0xff, 0xfd,
    0, 0, 0, 4, 0xff, 0xff, 0xff, 0xfb, 0, 0, 0, 6};
  const auto record = decode_record(payload.data(), payload.size());
  EXPECT_EQ(record.rdt_sequence, 0xffffffffU);
  EXPECT_EQ(record.ft_sequence, 0x80000000U);
  EXPECT_EQ(record.status, 0x80020000U);
  EXPECT_EQ(record.fx, -1); EXPECT_EQ(record.fy, 2); EXPECT_EQ(record.fz, -3);
  EXPECT_EQ(record.tx, 4); EXPECT_EQ(record.ty, -5); EXPECT_EQ(record.tz, 6);
}

TEST(Protocol, DecodesTwoComplementAxisBoundariesWithoutNarrowing)
{
  const std::array<std::uint8_t, 36> payload{
    0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 3,
    0x80, 0, 0, 0, 0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
    0, 0, 0, 0, 0x80, 0, 0, 1, 0x7f, 0xff, 0xff, 0xfe};
  const auto record = decode_record(payload.data(), payload.size());
  EXPECT_EQ(record.fx, std::numeric_limits<std::int32_t>::min());
  EXPECT_EQ(record.fy, std::numeric_limits<std::int32_t>::max());
  EXPECT_EQ(record.fz, -1);
  EXPECT_EQ(record.tx, 0);
  EXPECT_EQ(record.ty, std::numeric_limits<std::int32_t>::min() + 1);
  EXPECT_EQ(record.tz, std::numeric_limits<std::int32_t>::max() - 1);
}

TEST(Protocol, DoesNotUseImplementationDefinedUnsignedToSignedConversion)
{
  std::ifstream source{NETFT_PROTOCOL_SOURCE_PATH};
  const std::string contents{std::istreambuf_iterator<char>{source}, {}};
  EXPECT_EQ(contents.find("static_cast<std::int32_t>(read_u32"), std::string::npos);
}

class InvalidDatagramSize : public ::testing::TestWithParam<std::size_t> {};
TEST_P(InvalidDatagramSize, RejectsEveryNonRdtLength)
{
  const std::vector<std::uint8_t> payload(GetParam());
  try {
    static_cast<void>(decode_record(payload.data(), payload.size()));
    FAIL() << "malformed payload was accepted";
  } catch (const ProtocolError & error) {
    EXPECT_NE(std::string{error.what()}.find("exactly 36 bytes"), std::string::npos);
  }
}
INSTANTIATE_TEST_SUITE_P(Protocol, InvalidDatagramSize, ::testing::Values(0, 7, 35, 37, 72));

}  // namespace
